### PR TITLE
INT-3916: Don't use CTOR injection in FactoryBean

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3916

The `JpaOutboundGatewayFactoryBean` used CTOR injection for the `JpaExecutor`.
That one, in turn, uses CTOR injection for the `EntityManagerFactory`.

Such a dependency may cause the `early bean instantiating` in case of `AbstractAutowireCapableBeanFactory.getSingletonFactoryBeanForTypeCheck()`.
And we end up with the `BeanCurrentlyInCreationException`.

Therefore no one `FactoryBean` should use CTOR injection if there is a potential hierarchical dependency.

NOTE: there is no tests on the matter, since we don't change the components behavior.
The `JPA` sample application will be changed to the Boot to track this fix.

**Cherry-pick to 4.2.x**

### CODE REVIEW

File: spring-integration-jpa/src/main/java/org/springframework/integration/jpa/outbound/JpaOutboundGatewayFactoryBean.java [hunk 1]
```diff
@@ -71,18 +70,11 @@
...
 
-	/**
-	 * Constructor taking an {@link JpaExecutor} that wraps all JPA Operations.
-	 *
-	 * @param jpaExecutor Must not be null
-	 *
-	 */
-	public JpaOutboundGatewayFactoryBean(JpaExecutor jpaExecutor) {
-		Assert.notNull(jpaExecutor, "jpaExecutor must not be null.");
+	public void setJpaExecutor(JpaExecutor jpaExecutor) {
```
Comment (garyrussell) [inline spring-integration-jpa/src/main/java/org/springframework/integration/jpa/outbound/JpaOutboundGatewayFactoryBean.java]: I don't think we can remove the ctor if we're going to cherry-pick the change.

The problem is solved by the parser not using the ctor.

Maybe deprecate it?

**Comment by artembilan** [comment]
Pushed.

**Comment by garyrussell** [comment]
Minor polishing here https://github.com/garyrussell/spring-integration/commit/66144d865c126021e9bdb9886032f86e2abbf5bf

Methods called from ctors must be `final`. Since these are simple setters, it's easier to just set the field directly in the ctor.

What about the other FBs mentioned in the JIRA? (particularly those around Aggregator - where we saw the original problem with the `AggregatorFactoryBean`).

**Comment by artembilan** [comment]
Doh, forgot to comment about all other `FB`s.

Most of them are based on the `String`. `CorrelationStrategyFactoryBean` and `ReleaseStrategyFactoryBean` are out of use already in the Framework. And I guess it is just because of `AggregatorFactoryBean`. So, we even can deprecate them altogether.

Only one remains there - `CuratorFrameworkFactoryBean(String connectionString, RetryPolicy retryPolicy)` where we may worry only about the `RetryPolicy` bean injection.

**Comment by artembilan** [comment]
Thank you for letting to know about `final` for methods from ctors.

**Comment by garyrussell** [comment]
Interesting; as far as I can tell, those FBs have __never__ been used: https://github.com/spring-projects/spring-integration/commit/a5c14782fbd066feea80aaec9a2786d7617ebaad

**Comment by artembilan** [comment]
Sorry, Gary, we use them for Correlation Handler Parsers:
```
private static BeanMetadataElement createAdapter(BeanMetadataElement ref, String method, String unqualifiedClassName) {
		BeanDefinitionBuilder builder = BeanDefinitionBuilder
				.genericBeanDefinition(IntegrationConfigUtils.BASE_PACKAGE + ".config." + unqualifiedClassName
						+ "FactoryBean");
```
So, looks like we really should fix them as well.
Or continue to live with them as is until the urgent demand...

**Comment by garyrussell** [comment]
Ah - good - yes; we should fix them too.

**Comment by artembilan** [comment]
Pushed.

**Comment by garyrussell** [comment]
Stomp and WebSocket issues...

>Invalid property 'sockJsTaskScheduler' of bean class [org.springframework.integration.websocket.config.WebSocketIntegrationConfigurationInitializer$WebSocketHandlerMappingFactoryBean]: Bean property 'sockJsTaskScheduler' is not writable or has an invalid setter method. Does the parameter type of the setter match the return type of the getter?
	at org.springframework.beans.BeanWrapperImpl.createNotWritablePropertyException(BeanWrapperImpl.java:231)

**Comment by garyrussell** [comment]
Branch wasn't up to date.

**Comment by artembilan** [comment]
> Branch wasn't up to date.

There is really unfixed issue there in the websocket stuff.

Thank you for catching and sorry that I haven't tested it properly before pushing. 
Only `core` and `jpa` :smile:

**Comment by garyrussell** [comment]
Yes; still seeing problems in stomp

**Comment by artembilan** [comment]
Pushed.
```
gradlew api testAll --parallel
.....
BUILD SUCCESSFUL
               
Total time: 9 mins 38.438 secs
```

**Comment by garyrussell** [comment]
Merged as fec2a36f429ee5e2d407fecfdb23474df6c9b76f

Does not cherry-pick cleanly.

**Comment by artembilan** [comment]
> Does not cherry-pick cleanly.

Let me prepare backport PR!

**Pull request closed without merging**